tokio/src/runtime/blocking/sharded_queue.rs

```diff
@@ -0,0 +1,227 @@
+//! A sharded concurrent queue for the blocking pool.
+//!
+//! This implementation distributes tasks across multiple shards to reduce
+//! lock contention when many threads are spawning blocking tasks concurrently.
+//! The push operations use per-shard locking, while notifications use a global
+//! condvar for simplicity.
+//!
+//! The queue adapts to the current concurrency level by using fewer shards
+//! when there are few threads, which improves cache locality and reduces
+//! lock contention on the active shards.
+
+use crate::loom::sync::{Condvar, Mutex};
+
+use std::collections::VecDeque;
+use std::sync::atomic::Ordering::{AcqRel, Acquire, Relaxed, Release};
+use std::sync::atomic::{AtomicBool, AtomicUsize};
```

Comment:
Could we use loom's sync primitive and add a small test for this data structure? That would probably be something like [this](https://github.com/tokio-rs/tokio/blob/97d06ae1a69a38dfe20e65b5972f456781f2e0c1/tokio/src/runtime/tests/loom_multi_thread/queue.rs).